src/main/scala/monix/kafka/KafkaConsumer.scala

```diff
@@ -0,0 +1,140 @@
+
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+import scala.util.{Failure, Success}
+import scala.collection.JavaConverters._
+
+/** An `Observable` implementation that reads messages from `Kafka`.
+  *
+  * You need to call `addTopicSubscription` in order for the Observable to start sending messages from that topic.
+  */
+final class KafkaConsumer[K, V] private
+  (config: KafkaConsumerConfig, ioScheduler: Scheduler)
+  (implicit K: Deserializer[K], V: Deserializer[V])
+  extends Observable[ConsumerRecord[K, V]] with StrictLogging {
+
+  def unsafeSubscribeFn(subscriber: Subscriber[ConsumerRecord[K, V]]): Cancelable = {
+    val source = underlying.onErrorHandleWith { ex =>
+      logger.error(
+        s"Unexpected error in KafkaConsumerObservable, retrying in ${config.retryBackoffTime}", ex)
+      // Applying configured delay before retry!
+      underlying.delaySubscription(config.retryBackoffTime)
+    }
+
+    source.unsafeSubscribeFn(subscriber)
+  }
+
+  def addTopicSubscription(topics: Set[String]): Unit = {
+    logger.info(s"Subscribing to topics : ${topics.mkString(",")}")
+    consumerRef.subscribe((currentSubscriptions() ++ topics).asJavaCollection)
+  }
+
+  /**
+    * Removes a subscription from a particular topic.
+    *
+    * @param topics the topics to be unsubscribed, if it is an `Set.empty` this method will unsubscribe from all topics.
+    */
+  def removeTopicSubscription(topics: Set[String]): Unit =
+    if(topics.nonEmpty) {
+      logger.info(s"Unsubscribing from topics : ${topics.mkString(",")}")
+      consumerRef.subscribe((currentSubscriptions() -- topics).asJavaCollection)
+    } else {
+      logger.info(s"Unsubscribing from topics : ${currentSubscriptions().mkString(",")}")
+      consumerRef.unsubscribe()
+    }
+
+  // Gets initialized on `subscribe`
+  private lazy val consumerRef = {
+    logger.info(s"Kafka consumer connecting to servers: ${config.servers.mkString(",")}")
+    new ApacheKafkaConsumer[K,V](config.toProperties, K.create(), V.create())
+  }
+
+  private def currentSubscriptions() =
+    consumerRef.subscription().asScala.toSet
+
+  private lazy val underlying =
+    Observable.unsafeCreate[ConsumerRecord[K, V]] { subscriber =>
+      val cancelable = SingleAssignmentCancelable()
+
+      // Forced asynchronous boundary, in order to not block
+      // the current thread!
+      ioScheduler.executeAsync {
+        cancelable := Observable
+          .fromStateAction[Set[String], ConsumerRecords[K, V]]{ subscriptions =>
+            val pollRecords =
+              if (subscriptions.nonEmpty)
+                consumerRef.poll(config.fetchMaxWaitTime.toMillis)
+              else
+                ConsumerRecords.empty[K, V]()
+
+            (pollRecords, currentSubscriptions())
+          }(initialState = currentSubscriptions())
+          .doOnSubscriptionCancel(consumerRef.close())
```

Comment:
This `doOnSubscriptionCancel` seems convenient but I'd recommend against it, because:

1. the call to cancel is going to be concurrent with consuming within that observable, which means that when you do `consumer.close()` you have to `synchronize` it and set your `isDone = true`
2. you also need to do `consumer.close()` on `onComplete` and `onError`
3. you need to do error handling on `computer.close()` - basically **every Java call** needs to be protected against thrown exceptions

The fix is not hard though - if you stop having that `consumerRef` as a local that is.